lib/async_hooks.js

```diff
@@ -130,6 +131,60 @@ function createHook(fns) {
   return new AsyncHook(fns);
 }
 
+// AsyncLocal API //
+
+const locals = new Set();
+const localsHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    const execRes = executionAsyncResource();
+    locals.forEach((local) => {
+      local[kPropagateSymbol](execRes, resource);
+    });
+  }
+});
+
+const kResToValSymbol = Symbol('resToVal');
+const kPropagateSymbol = Symbol('propagate');
+
+class AsyncLocal {
+  constructor() {
+    this[kResToValSymbol] = new WeakMap();
+    locals.add(this);
+    localsHook.enable();
+  }
+
+  [kPropagateSymbol](execRes, initRes) {
+    const value = this[kResToValSymbol].get(execRes);
+    if (value !== undefined) {
+      this[kResToValSymbol].set(initRes, value);
+    }
+  }
+
+  get() {
+    if (this[kResToValSymbol]) {
+      return this[kResToValSymbol].get(executionAsyncResource());
+    }
+    return undefined;
+  }
+
+  set(value) {
+    if (!this[kResToValSymbol]) {
+      throw new ERR_ASYNC_LOCAL_CANNOT_SET_VALUE();
+    }
+    this[kResToValSymbol].set(executionAsyncResource(), value);
```

Comment:
This would set a barrier between the current execution layer and its parent, discarding the upper value and only propagating the new one downward through the tree. That's not necessarily a bad thing, but a user could very easily make the mistake of calling `set` again somewhere further into the request thinking it would allow a different path through the request to access the value, which is not the case. In my experience, it's better for the barrier to be applied in the constructor that way the object can't unexpectedly fork part-way through a request. Rather than constructing _one_ `AsyncLocal` instance at the top-level, you'd construct a new one at the start of each request.

I know there's also the use-case of enabling isolated components of a system to share data without explicit communication, and I think the _best_ way to achieve that is just using a _factory_ that can be called to produce or retrieve `AsyncLocal` instances. In a way, you _kind of_ already have this with the `set` and `get`, but I feel the naming is misleading and should be made clearer that `set` is _forking_ the following async tree to use the new context object.

Reply:
> Rather than constructing one AsyncLocal instance at the top-level, you'd construct a new one at the start of each request.

With this API such usage should be considered a bad practice, because of performance reasons and tricky API manipulations. By the latter I mean that it's hard to share individual `AsyncLocal` instance with each request. It's much simpler to use a single instance to handle all requests. See this example: https://github.com/nodejs/node/pull/31016/files#diff-9a4649f1c3f167b0da2c95fc38953a1fR485

As for merging values or doing any other actions on value modification, that could (and, in my opinion, should) be done in user-land, i.e. in application code or libraries, because exact logic depends on particular use case.

> In a way, you kind of already have this with the set and get, but I feel the naming is misleading and should be made clearer that set is forking the following async tree to use the new context object.

I think that `.get()`/`.set(value)` method names are quite intuitive for anyone who is familiar with the concept of asynchronous call chain. To help users, PR adds a couple of examples showing how values are propagated into the documentation.

For me, alternative method names, like `.read()`/`.fork(value)` or something like that are much less obvious.

Reply:
I feel the opposite. The `.get()` and `.set(value)` names are not intuitive _at all_ to someone trying to use the API to solve the need of having a unique context object per-request. It's capable of _achieving_ that by using `.set(value)` at the start of the request to store an object and then _only_ using `.get()` to retrieve and modify that object for the rest of the request, but that's _very_ non-obvious from the naming or the docs. The only reason it makes sense to _us_ is because we are already intimately familiar with the concept of async call trees and how this feature is implemented. An average user is not, and I'm almost 100% certain users will try to use this like ThreadLocal in Java and then be _totally confused_ when different parts of their app get different data because they keep calling `.set(value)` to "update" the stored value; it looks _a lot_ like `setState(...)` in React, and I suspect many people will treat it as such. I'm sure many users will also try to construct a new AsyncLocal within each request as that's a common practice with ThreadLocal in Java, again doing it the _wrong_ way here.

In my opinion, a proper async context API should look something more like this:

```js
// This corresponds roughly to the existing `AsyncLocal`
const contexts = new ContextManager()

function loadUser(auth) {
  const { onError, onUser } = contexts.current
  User.login(auth).then(onUser, onError)
}

function httpResponder(res) {
  contexts.current.onError = err => {
    res.status(500).end(err.message)
  }
  contexts.current.onUser = user => {
    res.status(200).end(user.token)
  }
}

app.post('/login', (req, res) => {
  // This corresponds to `.set(value)` and is closest
  // to what most people actually think of as a local.
  contexts.create()
  httpResponder(res)
  loadUser(req.headers.authorization)
})
```

It's functionally identical to what you have built, but more cleanly communicates the intent of the objects and methods.

Reply:
The snippet looks like something that can be built on top of `AsyncLocal` API, but personally I find it less straightforward. `AsyncLocal` API, in its current state, allows building APIs on top of it, say, the one from the snippet or something similar to `cls-hooked`/`continuation-local-storage` API. So, I expect it to be mainly used by library authors, not all node users.

As for user confusion, I hope that it can be mitigated by good enough documentation.